app/src/org/commcare/dalvik/application/CommCareApplication.java

```diff
@@ -860,9 +861,22 @@ public void onServiceConnected(ComponentName className, IBinder service) {
 
                     mBoundService = ((CommCareSessionService.LocalBinder)service).getService();
 
-                    //Don't let anyone touch this until it's logged in
+                    // Don't let anyone touch this until it's logged in
                     // Open user database
-                    mBoundService.prepareStorage(key, record);
+                    try {
+                        mBoundService.prepareStorage(key, record);
+                    } catch (MigrationException e) {
+                        if (e.isDefiniteFailure()) {
+                            CommCareApplication._().triggerHandledAppExit(CommCareApplication.this,

```

Comment:
This strategy still seems unfortunate to me, since there's no action a user can take that will set things back to a reasonable state. This will just keep crashing the app over and over, even though there are three things we can potentially do:

1) Clear the current user data and log that user in again
2) Clear app data and re-install the app (but not uninstall all apps)
3) Something smart per migration, like moving the forms somewhere that they can be submitted.

For this app, for instance, if people only had Saved/Incomplete forms they're basically fine wiping all of them out. It's only unsent forms that are a problem, right?